README.md
# Computation Against a Neighbour

This is the project repository for the case study
 presented in a homonym manuscript submitted to a journal.

## Structure of the repository

The project has the standard structure of Maven/Gradle projects.

* `./build.gradle` contains the build definition and the tasks for compiling and packaging the application
* `./src/main/scala/it/unibo/casestudy/AdHocCloud.scala` contains the **source code of the case study**
* `./src/main/yaml/adhoccloud.yml` contains the **configuration of the Alchemist simulation**
* `./plotter.py` is a script that can be used to plot data produced by Alchemist
* `./history/` contains, for different simulations, the corresponding configuration files

## How-to

A simulation can be started with a command as the following:

```bash
$ ./gradlew --no-daemon && \
  java -Xmx5024m -cp "build/classes/main:libs/alchemist-redist-7.0.2-beta.jar:libs/scafi-lib-assembly-0.3.0.jar:build/resources/main" \
  it.unibo.alchemist.Alchemist \
  -b -var random leaderFailureProbability \
  -y src/main/yaml/adhoccloud.yml -e data/20181010_exp -t 400 -p 3 -v &> exec.txt &
```

where

* `./gradlew --no-daemon` is used to compile and build the simulation source code
* `java -cp "..." it.unibo.alchemist.Alchemist` sets the java classpath and the main class to run
* `-b -var <var1> <var2>` sets batch mode, and the given variables are combined to form a matrix of simulation configurations to run
* `-y <file>` specifies the configuration for the simulation to run (which describes variables, i.e., input parameters, and exports)
- `-e <basefilepath>` specifies the base filepath for generating data output files
- `-t 400` specifies the logical time to be simulated; `-p <N>` specifies the number of parallel simulations to run;
- `-v &> exec.txt &` sets the logging level as verbose and redirect output to a text file (e.g., for debugging purposes) and starts the task in background

Once data files (which are in CSV format) are generated, you can use script `plotter.py` to create plots.

```bash
# ./plotter.py <dir> <baseFilepath> <plotConfigFile>
$ ./plotter.py data 20181010_exp plotconfig.yml
```

where `plotconfig.yml` is something as the following:

```yaml
the_plots_labels: # these should map the 'export fields'
  - Time
  - Available CPU
  - Actual CPU usage
  - Estimated CPU usage
  - Nfailures
  - Nleaders
the_plots_formats: # the plots to be created, and which fields to be selected
  - [0,2,3,4,5]    # the first field is used for X-axis; the others appear as curves
line_widths:
  - [1,1,1,1,1,1]
default_colors: &default_colors ["black","blue", "red","darkgreen","orange","violet"]
the_plots_colors:
  - *default_colors
legend_position:
  - "lower right"
sampling: True # when true, you use a 'random' (batch) dimension to unify multiple files and averaging the curves

```

## Reproducing the experiments

NOTE: this may take a few hours.

In the paper, Fig. 12 shows, for different `leaderProbabilityFailure` values, the estimated CPU usage
 against the actual CPU usage, taking the mean values of 30 runs with varying random seeds.
These plots can be re-generated with the following commands:

```bash
$ git fetch --all --tags --prune
$ git checkout tags/fig12 -b fig12

$ ./gradlew --no-daemon && \
 java -Xmx5024m -cp "build/classes/main:libs/alchemist-redist-7.0.2-beta.jar:libs/scafi-lib-assembly-0.3.0.jar:build/resources/main" \
 it.unibo.alchemist.Alchemist \
 -b -var leaderFailureProbability random \
 -y history/fig12/adhoccloud.yml \
 -e data/fig12 -t 850 -p 3 -v &> exec.txt &

$ ./plotter.py data fig12 history/fig12/plot.yml
```

## Contacts

* Roberto Casadei: roby [dot] casadei [at] unibo [dot] it

## Links

* [ScaFi repository](https://github.com/scafi/scafi)
* [ScaFi website](https://scafi.github.io/)


history/history.txt
./gradlew --no-daemon && java -Xmx5024m -cp "build/classes/main:libs/alchemist-redist-7.0.2-beta.jar:libs/scafi-lib-assembly-0.3.0.jar:build/resources/main" it.unibo.alchemist.Alchemist        -b -var random leaderFailureProbability         -y src/main/yaml/adhoccloud.yml -e data/20181011_fexp -t 1000 -p 3 -v &> exec.txt &
./plotter.py data 20181011_fexp 20181011_fexp.yml

./gradlew --no-daemon && java -Xmx5024m -cp "build/classes/main:libs/alchemist-redist-7.0.2-beta.jar:libs/scafi-lib-assembly-0.3.0.jar:build/resources/main" it.unibo.alchemist.Alchemist -b -var random -y src/main/yaml/adhoccloud.yml -e data/20181012_upmetric -t 850 -p 3 -v &> exec.txt
./plotter.py data 20181012_upmetric 20181012_upmetric.yml


src/main/scala/it/unibo/casestudy/AdhocCloud.scala
package it.unibo.casestudy

import it.unibo.ScafiAlchemistSupport
import it.unibo.alchemist.model.implementations.positions.LatLongPosition
import it.unibo.alchemist.model.scafi.ScafiIncarnationForAlchemist._
import it.unibo.scafi.core.Core
import it.unibo.scafi.space.Point3D

trait DiskType
case object HDD extends DiskType
case object SSD extends DiskType

case class Percent(p: Double) {
  //assert(p>=0 && p<=100)

  def of(value: Double): Double = value/100.0*p
}
object Percent {
  implicit def toDouble(p: Percent): Double = p.p
  def of(v1: Double, v2: Double): Percent = Percent(v2*100/v1)
}

case class ProcessingStat(numCPUs: Int = 0, mHz: Int = 0, numCPUsInUse: Int = 0) {
  val inUse = Percent.of(numCPUs,numCPUsInUse)
}
case class MemoryStat(MiBs: Int = 0, MiBsInUse: Int = 0){
  val inUse = Percent.of(MiBs, MiBsInUse)
}
case class StorageStat(MiBs: Int = 0, dtype: DiskType = HDD, MiBsInUse: Int = 0) {
  val inUse = Percent.of(MiBs, MiBsInUse)
}
case class LocalStats(processing: ProcessingStat = ProcessingStat(),
                      memory: MemoryStat = MemoryStat(),
                      storage: StorageStat = StorageStat()){
  val inUse = Percent((processing.inUse.p + memory.inUse.p + storage.inUse.p)/3)
}
case class AggregateStats(cpus: Int = 0, memoryMiBs: Int = 0, storageMiBs: Int = 0,
                          cpusInUse: Long = 0, memoryMiBsInUse: Long = 0, storageMiBsInUse: Long = 0){
  def aggregate(ls: LocalStats) = AggregateStats(
    cpus = this.cpus + ls.processing.numCPUs,
    memoryMiBs = this.memoryMiBs + ls.memory.MiBs,
    storageMiBs = this.storageMiBs + ls.storage.MiBs,
    cpusInUse = this.cpusInUse + ls.processing.numCPUsInUse,
    memoryMiBsInUse = this.memoryMiBsInUse + ls.memory.MiBsInUse,
    storageMiBsInUse = this.storageMiBsInUse + ls.storage.MiBsInUse,
  )
}
case class AreaStats(leader: Int)(val stats: AggregateStats){
  override def toString: String = s"($leader, $stats)"
}
case class SpatialData(leader: Int, leaderPosition: LatLongPosition, meanDensity: Double)

// The function requires an Aggregate Program to avoid issues related to passing functions
// belonging to different AggregatePrograms.
case class Fun[R](ver: Int, fun: (AggregateProgram with StandardSensors)=>R)

case class Gradient(algorithm: (Boolean,()=>Double)=>Double, source: Boolean, metric:()=>Double){
  def run(): Double = algorithm(source, metric)
}

case class RequestMoreResources(leader: Int)(val position: LatLongPosition)

class AdhocCloud extends AggregateProgram
  with StandardSensors with ScafiAlchemistSupport with BlockG with BlockC with FieldUtils { self =>

  val RequestHandled = "request_handled"
  val IsMoving = "is_moving"
  val MoveTo = "move_to"

  override type MainResult = Any
  type Stats = Map[ID,LocalStats]

  lazy val idle =
    if(getPosition().y>44.1379 && nextRandom>0.38 ) true else if(nextRandom>0.99) true else false
  lazy val exhausted =
    mid%3!=0

  lazy val changeEveryMaxRounds =
    env.get[Int]("changeEveryMaxRounds")

  def localProcessors(initRandom: Double, dynamicRandom: Double) = {
    val ncpus = if(initRandom>0.8) 8 else if(initRandom>0.6) 4 else if(initRandom > 0.4) 2 else 1
    ProcessingStat(
      numCPUs = ncpus,
      mHz = (1 + Math.round(initRandom * 1).toInt)*1000,
      numCPUsInUse = /*if(idle) 0 else if(exhausted) ncpus else*/ (dynamicRandom * ncpus).toInt)
    // sense[ProcessingStat]("statsCPU")
  }
  def localMemory (initRandom: Double, dynamicRandom: Double) = {
    val mibs = 1 + (initRandom * 10).toInt*512
    MemoryStat(
      MiBs = mibs,
      MiBsInUse = if(idle) 0 else if(exhausted) (mibs/100.0*90).toInt else (dynamicRandom * mibs).toInt)
    // sense[MemoryStat]("statsRAM")
  }
  def localStorage(initRandom: Double, dynamicRandom: Double) = {
    val mibs = (1 + Math.round(initRandom * 50).toInt*2)*512
    StorageStat(MiBs = mibs,
      dtype = if(dynamicRandom>0.8) SSD else HDD,
      MiBsInUse = if(idle) 0 else (dynamicRandom * mibs).toInt)
    // sense[StorageStat]("statsStorage")
  }

  def metricInjecter(k: Int): Injecter[Double] = () =>
    if(mid != 0 || k < metricInjectedAt){ Fun(0, ap => ap.nbrRange()) } else { Fun(1, ap => ap.nbrRange()+env.get[Double]("metricDelta")) }

  lazy val metricInjectedAt: Int = env.get[Int]("metricInjectedAt")
  lazy val lowDensity: Double = env.get[Double]("lowDensityThreshold")
  lazy val highDensity: Double = env.get[Double]("highDensityThreshold")
  lazy val lowLoad: Double = env.get[Double]("lowLoad")
  lazy val highLoad: Double = env.get[Double]("highLoad")

  def getPosition(): LatLongPosition =
    sense[LatLongPosition]("position")

  implicit def fromLatLogToPoint(ll: LatLongPosition): Point3D =
    Point3D(ll.getLongitude, ll.getLatitude, 0)

  val IsLeader = "leader"

  lazy val initialization: Unit = {
    env.put(RequestHandled, Option.empty[RequestMoreResources])
    env.put(IsMoving, false)
    env.put("distance", 0.0)
    env.put(MoveTo, new LatLongPosition(0,0))
    env.put(IsLeader, false)
    env.put("failed", 0)
  }

  def keepTrackOfDistance() =     rep(getPosition()){ case pos =>
    env.put("distance", env.get[Double]("distance") + getPosition().getDistanceTo(pos))
    getPosition()
  }

  override def main(): Any = {
    initialization
    val k = rep(0)(_+1)
    val initRandom = rep(nextRandom){ r => r }
    val dynamicRandom = rep((20+(nextRandom*changeEveryMaxRounds).toInt, nextRandom)) { case (step,rand) =>
      //env.put("dynamic_random", s"$k => ${k%step} => Every $step changes to $rand")
      if(k % step == 0) (step,nextRandom + (if((k>200 && k<400) || (k>500 && k <600)) 0.4 else 0)) else (step,rand)}._2

    val localResources = LocalStats(localProcessors(initRandom, dynamicRandom), localMemory(initRandom, dynamicRandom), localStorage(initRandom, dynamicRandom))

    env.put("this",self.hashCode())
    env.put("this_position", getPosition())
    env.put("local_resources", localResources)
    env.put("available_cpu", localResources.processing.numCPUs)
    env.put("usage_cpu", localResources.processing.numCPUsInUse)
    env.put("usage_memory", localResources.memory.MiBsInUse)
    env.put("usage_storage", localResources.storage.MiBsInUse)
    env.put("usage", (localResources.processing.inUse.p + localResources.memory.inUse.p +localResources.storage.inUse.p)/3.0)
    // keepTrackOfDistance()

    val metricfun: Fun[Double] = up[Double](metricInjecter(k))
    env.put("metric", metricfun)
    val metric: () => Double = () => metricfun.fun(this) // nbrRange // NB: trick to avoid issues related to shipping lambdas from other classes

    val leaders = branch(env.get[Boolean](IsLeader) & nextRandom<env.get[Double]("failureProb")){
      env.put("failed", env.get[Int]("failed")+1)
      false
    }{
      S(env.get[Double]("grain"), metric)
    }
    val potential = distTo(leaders) // metric

    val resourcesPerArea = C[Double,Stats](potential, _++_, Map(mid -> localResources), Map())
    val aggregatedResourcesPerArea = resourcesPerArea.values.foldLeft(AggregateStats())(_.aggregate(_))
    val allResourcesInArea = broadcast(potential, metric, aggregatedResourcesPerArea)

    val numNodesInArea = C[Double,Int](potential, _+_, 1, 0)
    val numNbrs = excludingSelf.sumHood(1)
    val numNbrsInArea = C[Double,Double](potential, _+_, numNbrs, 0)
    val spatialDataForArea = broadcast(potential, metric, SpatialData(mid, getPosition(), numNbrsInArea/numNodesInArea))
    val meanDensity = spatialDataForArea.meanDensity

    val myLeader = broadcast(potential, metric, mid)
    val nbrAreas = excludingSelf.unionHood(nbr{AreaStats(myLeader)(allResourcesInArea)})

    env.put("estimated_usage_cpu", if(leaders) aggregatedResourcesPerArea.cpusInUse else 0)

    val adjacentAreasStats: Set[AreaStats] = broadcast(potential, metric, C[Double,Set[AreaStats]](potential, _.union(_), nbrAreas, Set()))

    val moreDevicesAllowed =
      meanDensity < highDensity
    val moreCPUNeeded =
      Percent.of(allResourcesInArea.cpus, allResourcesInArea.cpusInUse) > highLoad
    val mayMove = !leaders &&
      localResources.inUse.p == 0 &&
      meanDensity > lowDensity &&
      Percent.of(allResourcesInArea.cpus, allResourcesInArea.cpusInUse) < lowLoad &&
      sufficientlyStable(myLeader, 100)

    env.put("working_on_request", requestBeingHandled)
    branch(requestBeingHandled.isDefined){
      if(myLeader == requestBeingHandled.get.leader) doFinishRequest
    }{
      val requests = excludingSelf.unionHoodSet(nbr { mux(moreCPUNeeded){
        Set(RequestMoreResources(spatialDataForArea.leader)(spatialDataForArea.leaderPosition))
      }{ Set.empty } }).filter(_.leader!=myLeader)
      env.put("requests", requests)
      // Select one request
      if(mayMove && !requests.isEmpty) doHandleRequest(requests.head)
    }

    env.put("may_move", mayMove)
    env.put("may_move_str", s"$mayMove --- !busy (${localResources.inUse}) " +
      s" && !lowDensity ($meanDensity > $lowDensity = ${meanDensity > lowDensity}) && lowLoad (${Percent.of(allResourcesInArea.cpus, allResourcesInArea.cpusInUse)} < ${lowLoad})")
    env.put("more_devices_allowed", s"${meanDensity} < ${highDensity} = $moreDevicesAllowed")
    env.put("more_cpu_needed", s"${Percent.of(allResourcesInArea.cpus, allResourcesInArea.cpusInUse)} > ${highLoad} = $moreCPUNeeded")
    env.put("mean_density", meanDensity)
    env.put(IsLeader, leaders)
    env.put("adjacent_area_stats", adjacentAreasStats)
    env.put("numNodes", if(leaders) numNodesInArea else 0)
    env.put("aggregateResourcesPerArea", aggregatedResourcesPerArea)
    env.put("allResourcesPerArea", allResourcesInArea)
    env.put("myLeader", myLeader)
    env.put("potential", potential)

    77
  }

  def doHandleRequest(req: RequestMoreResources) = { env.put(RequestHandled, Some(req)); env.put(IsMoving, true); env.put(MoveTo, req.position) }
  def doFinishRequest = { env.put(RequestHandled, None); env.put(IsMoving, false); env.put(MoveTo, getPosition()) }
  def isWorkingOnRequest = env.get(RequestHandled)!=None
  def requestBeingHandled: Option[RequestMoreResources] = if(isWorkingOnRequest) env.get(RequestHandled) else Option.empty

  def captureChange[T](x: T): Boolean = rep((x,false)) { case (value, _) =>
    (x, value != x)
  }._2

  def sufficientlyStable[T](x: T, k: Int) = rep((x,0,false)){ case (v,nrounds,suffStable) =>
      mux(v!=x){ (x, 0, false) }{ (x, nrounds+1, nrounds+1 >= k) }
  }._3

  def classicGradient(source: Boolean, metric: () => Double = nbrRange) = rep(Double.PositiveInfinity){ case d =>
    mux(source){ 0.0 }{ minHoodPlus(nbr(d)+metric()) }
  }

  def distTo(source: Boolean, metric: () => Double = nbrRange) =
    classicGradient(source, metric)

  def G[V](gradient: Gradient, field: V, acc: V => V): V = {
    val g = gradient.run()
    rep(field) { case (value) =>
      mux(g==0.0){ field }{ excludingSelf.minHoodSelector[Double,V](nbr{g}+gradient.metric())(acc(nbr{value})).getOrElse(field) }
    }
  }

  def G[V](g: Double, metric: ()=>Double, field: V, acc: V => V): V = {
    rep(field) { case (value) =>
      mux(g==0.0){ field }{ excludingSelf.minHoodSelector[Double,V](nbr{g}+metric())(acc(nbr{value})).getOrElse(field) }
    }
  }

  def broadcast[V](source: Boolean, field: V): V =
    G[V](Gradient(classicGradient,source, nbrRange), field, (v: V) => v)

  def broadcast[V](potential: Double, metric: () => Double, field: V): V =
    G(potential, metric, field, identity(_:V))

  /**
    * NOTE: the transitory phase is different with respect to the "classical gradient".
    * I.e., devices are at a "distance" 0 until accumulation of values along the gradient make their distance raise.
    */
  def distToWithG(source: Boolean, metric: () => Double = nbrRange): Double =
    G[Double](Gradient(classicGradient, source, metric), 0.0, (d: Double) => d+metric())

  def S(grain: Double, metric: () => Double): Boolean =
    breakUsingUids(randomUid, grain, metric)

  def randomUid: (Double, ID) =
    rep((nextRandom, mid())) { v => (v._1, mid()) }

  def breakUsingUids(uid: (Double, ID),
                     grain: Double,
                     metric: () => Double): Boolean =
    uid == rep(uid) { lead: (Double, ID) =>
      distanceCompetition(distTo(uid == lead, metric), lead, uid, grain, metric)
    }

  /**
    * Candidate leader devices surrender leadership to the lowest nearby UID.
    *
    * @return
    */
  def distanceCompetition(d: Double,
                          lead: (Double, ID),
                          uid: (Double, ID),
                          grain: Double,
                          metric: () => Double): (Double, ID) = {
    val inf: (Double, ID) = (Double.PositiveInfinity, uid._2)
    mux(d > grain) {
      uid
    } {
      mux(d >= (0.5 * grain)) {
        inf
      } {
        minHood {
          mux(nbr { d } + metric() >= 0.5 * grain) {
            nbr { inf }
          } {
            nbr { lead }
          }
        }
      }
    }
  }

  // UPDATABLE METRIC
  type Injecter[R] = () => Fun[R]
  def up[R](injecter: Injecter[R]): Fun[R] = rep(injecter()){ case f =>
    //val injfun = injecter()
    //val nbrfun = includingSelf.maxHoodSelector(nbr{ f.ver })(nbr{ f }).getOrElse(injfun)
    //if(injfun.ver > nbrfun.ver) injfun else nbrfun
    foldhood(injecter())((f1,f2) => if(f1.ver>=f2.ver) f1 else f2)(nbr{f})
  }

  implicit class RichFieldOps(fieldOps: FieldOps) {
    def maxHoodSelector[T: Builtins.Bounded, V](toMaximize: => T)(data: => V): Option[V] = {
      val ord = implicitly[Builtins.Bounded[T]]
      fieldOps.foldhoodTemplate[(T, Option[V])]((ord.bottom, None))((x, y) => if (ord.compare(x._1, y._1) > 0) x else y)((toMaximize, Some(data)))._2
    }
  }
}